### PR TITLE
Completed all waves and ran truffula command line

### PR DESCRIPTION
**Pull request by TiaMarieG**
The extra credit for wave 3 has also, potentially, been completed. We believe our code and test cases account for the described edge case.

### CODE REVIEW

File: src/ColorPrinter.java [hunk 1]
```diff
@@ -86,7 +86,15 @@ public void print(String message) {
...
   public void print(String message, boolean reset) {
-    // TODO: Implement this!
+    if (!reset){
+      printStream.append(currentColor + message);
+    }
+    
+    else {
+      printStream.append(currentColor + message + ConsoleColor.RESET);
+      currentColor = ConsoleColor.RESET;
+    }
```
Comment (auberonedu) [inline src/ColorPrinter.java:96]: Nice!

File: src/ColorPrinterTest.java [hunk 2]
```diff
@@ -26,4 +26,147 @@ void testPrintlnWithRedColorAndReset() {
...
+    // Arrange
+    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    PrintStream printStream = new PrintStream(outputStream);
+
+    ColorPrinter printer = new ColorPrinter(printStream);
+    printer.setCurrentColor(ConsoleColor.BLUE);
+
+    // Act: Print the message
+    String message = "I speak for the ocean";
+    printer.println(message);
```
Comment (auberonedu) [inline src/ColorPrinterTest.java:44]: 🌊

File: src/ColorPrinterTest.java [hunk 3]
```diff
@@ -26,4 +26,147 @@ void testPrintlnWithRedColorAndReset() {
...
+   ColorPrinter print = new ColorPrinter(printStream);
+   print.setCurrentColor(ConsoleColor.CYAN);
+   print.setCurrentColor(ConsoleColor.RESET);
+   print.setCurrentColor(ConsoleColor.PURPLE);
+ 
+   // Act
+   ConsoleColor currentColor = print.getCurrentColor();
+ 
+   // Assert
+   assertEquals(ConsoleColor.PURPLE, currentColor);
```
Comment (auberonedu) [inline src/ColorPrinterTest.java:149]: You might also consider printing in multiple colors and making sure the full output is printed correctly.

File: src/TruffulaOptions.java [hunk 4]
```diff
@@ -101,10 +101,49 @@ public boolean isUseColor() {
...
+        }
+        // TODO: handle a color argument like ["red", "orange", "yellow"]
+        // assign a new arraylist of given colors or empty array list if no given colors
+        else {
+          throw new IllegalArgumentException(currentString + "is not a valid flag");
+        }
+      }
+      else {
+        throw new FileNotFoundException(currentString + " is not found");
+      }
```
Comment (auberonedu) [inline src/TruffulaOptions.java:131]: Nice!

File: src/TruffulaOptions.java [hunk 5]
```diff
@@ -101,10 +101,49 @@ public boolean isUseColor() {
...
+    }
+
+    // check if file exists
+    if (!foundFile.exists()){
+      throw new FileNotFoundException(foundFile + " does not exist.");
+    }
+
+    root = foundFile;
+    showHidden = thisHidden;
+    useColor = thisColor;
```
Comment (auberonedu) [inline src/TruffulaOptions.java:146]: Consider using the instance variables directly instead of making local ones and then assignment them at the end.

File: src/TruffulaOptionsTest.java [hunk 6]
```diff
@@ -26,4 +28,68 @@ void testValidDirectoryIsSet(@TempDir File tempDir) throws FileNotFoundException
...
+      directory.mkdir();
+      String directoryPath = directory.getAbsolutePath();
+      
+      String[] args = {directoryPath};
+
+      directory.delete();
+
+      assertThrows(FileNotFoundException.class, () -> {
+        new TruffulaOptions(args);
+      });
```
Comment (auberonedu) [inline src/TruffulaOptionsTest.java:83]: Nice!

File: src/TruffulaPrinter.java [hunk 7]
```diff
@@ -76,43 +78,95 @@ public TruffulaPrinter(TruffulaOptions options, PrintStream outStream, List<Cons
...
-    out.println("My options are: " + options);
+    // base case
+    printTree(options.getRoot());
+  }
+
+  private void printTree(File root) {
+    // set color sequence
+    if (options.isUseColor()) {
+      // TODO: if given color array list is not empty -> assign that to color sequence, else use default color sequence
+      this.colorSequence = DEFAULT_COLOR_SEQUENCE;
```
Comment (auberonedu) [inline src/TruffulaPrinter.java:129]: This incorrectly overwrites the color sequence passed into the constructor. Also, `this` is not needed here.

File: src/TruffulaPrinter.java [hunk 8]
```diff
@@ -76,43 +78,95 @@ public TruffulaPrinter(TruffulaOptions options, PrintStream outStream, List<Cons
...
+
+    // Check for null children
+    if (root.listFiles() == null) {
+      return;
+    }
+
+    // traverse
+    for (var file : AlphabeticalFileSorter.sort(root.listFiles())) {
+      printTree(file, threeSpaces, colorIndex);
+    }
```
Comment (auberonedu) [inline src/TruffulaPrinter.java:169]: Nice recursion!

**Review by auberonedu — COMMENTED** [review]
Great job!
